core/lib/dal/src/blocks_dal.rs

```diff
@@ -549,14 +591,122 @@ impl BlocksDal<'_, '_> {
         Ok(storage_oracle_info.and_then(DbStorageOracleInfo::into_optional_batch_oracle_info))
     }
 
-    pub async fn set_eth_tx_id(
+    /// Get Last L2 blocks in batch with their rolling txs hash and precommit eth transactions ids,
+    /// grouped by batches.
+    pub async fn get_last_l2_block_rolling_txs_hashes_by_batches(
+        &mut self,
+    ) -> DalResult<HashMap<L1BatchNumber, Vec<L2BlockWithEthTx>>> {
+        let rows = sqlx::query!(
+            r#"
+            SELECT
+                l1_batches.number AS l1_batch_number,
+                miniblocks.eth_precommit_tx_id,
+                miniblocks.rolling_txs_hash AS "rolling_txs_hash!",
+                miniblocks.number AS miniblock_number
+            FROM
+                l1_batches
+            JOIN LATERAL (
+                SELECT
+                    eth_precommit_tx_id,
+                    rolling_txs_hash,
+                    number
+                FROM miniblocks
+                LEFT JOIN
+                    eth_txs_history
+                    ON eth_txs_history.eth_tx_id = miniblocks.eth_precommit_tx_id
+                WHERE
+                    l1_batch_number = l1_batches.number
+                    AND (
+                        eth_txs_history.finality_status != 'pending'
+                        OR eth_precommit_tx_id IS NULL
+                    )
+                ORDER BY number DESC
+                LIMIT 2
+            ) miniblocks ON TRUE
+            WHERE
+                l1_batches.number > 0
+                AND l1_batches.is_sealed
+                AND l1_batches.eth_commit_tx_id IS NULL
+                AND l1_batches.final_precommit_eth_tx_id IS NULL
+                AND miniblocks.rolling_txs_hash IS NOT NULL
```

Comment:
This query is expected to work fast because there is an index on `l1_batches.eth_commit_tx_id` so it will just scan rows where `l1_batches.eth_commit_tx_id IS NULL`. But please double-check it, maybe run this query against stage2 db and check its plan